envoy: Use separate clusters for egress and ingress redirects.

jrajahalme opened the pull request:
Currently it is possible that an ingress proxy accidentally reuses a
connection opened by an egress proxy, since in both cases the
connections share the same (original) destination address and the same
source security ID. Fix this by using separate clusters for ingress
and egress redirects, as each cluster has its own connection pool.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/5763)
<!-- Reviewable:end -->

rlenglet commented on pkg/envoy/server.go:
```diff
@@ -199,7 +204,7 @@ func StartXDSServer(stateDir string) *XDSServer {
 									}}}},
 									"route": {Kind: &structpb.Value_StructValue{StructValue: &structpb.Struct{Fields: map[string]*structpb.Value{
-										"cluster":          {Kind: &structpb.Value_StringValue{StringValue: "cluster1"}},
+										// "cluster":          {Kind: &structpb.Value_StringValue{StringValue: "cluster1"}},
```
Remove the commented line.

jrajahalme replied:
I've left the comments here on purpose as placeholders for members that are added for a specific redirect. They help reminding that the lines below that actually set these members need to be adjusted if the filters are moved around etc.

rlenglet commented on pkg/envoy/server.go:
```diff
@@ -231,7 +236,7 @@ func StartXDSServer(stateDir string) *XDSServer {
 			Config: &structpb.Struct{Fields: map[string]*structpb.Value{
 				"stat_prefix": {Kind: &structpb.Value_StringValue{StringValue: "tcp_proxy"}},
-				"cluster":     {Kind: &structpb.Value_StringValue{StringValue: "cluster1"}},
+				// "cluster":     {Kind: &structpb.Value_StringValue{StringValue: "cluster1"}},
```
Remove the commented line.

jrajahalme replied:
Same here.

jrajahalme committed:
envoy: Use separate clusters for egress and ingress redirects.

Currently it is possible that an ingress proxy accidentally reuses a
connection opened by an egress proxy, since in both cases the
connections share the same (original) destination address and the same
source security ID. Fix this by using separate clusters for ingress
and egress redirects, as each cluster has its own connection pool.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>
